onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -746,6 +746,65 @@ void TorchMultinomialKernelLauncher(float* d_input,
                                                 d_presence_mask);
 }
 
+__global__ void UpdateDecoderMaskedMultiheadAttentionCacheIndirectionKernel(int32_t* tgt_indir_cache,
+                                                                            const int32_t* src_indir_cache,
+                                                                            const int32_t* beam_ids,
+                                                                            int batch_size,
+                                                                            int beam_width,
+                                                                            int input_seq_length,
+                                                                            int max_seq_length,
+                                                                            int current_length) {
+  int time_step = threadIdx.x + blockIdx.x * blockDim.x;
+  int bb_id = threadIdx.y + blockIdx.y * blockDim.y;
+  const int batch_id = bb_id / beam_width;
+  const int beam_id = bb_id % beam_width;
+
+  if (bb_id >= beam_width * batch_size || time_step >= current_length) {
+    return;
+  }
+
+  const int src_beam = beam_ids[batch_id * beam_width + beam_id] % beam_width;
+
+  const int tgt_offset = batch_id * beam_width * max_seq_length + beam_id * max_seq_length + time_step;
+
+  if (time_step < input_seq_length) {
+    // For time steps that correspond to the input sequence,
+    // the beam that it comes from is always 0.
+    tgt_indir_cache[tgt_offset] = static_cast<int32_t>(0);
+  } else if (time_step == (current_length - 1)) {
+    // For the final (newly generated) time step,
+    // the beam that it comes from is always the beam that we
+    // are currently processing (i.e.) from this point on, these time-steps
+    // form the new beams.
+    tgt_indir_cache[tgt_offset] = static_cast<int32_t>(beam_id);
+  } else {
+    // For all other time-steps, we look up the source indirection, to
+    // see which beam it came from based on the `src_beam`.
+    const int src_offset = batch_id * beam_width * max_seq_length + src_beam * max_seq_length + time_step;
+    tgt_indir_cache[tgt_offset] = src_indir_cache[src_offset];
+  }
+}
+
+void UpdateDecoderMaskedMultiheadAttentionCacheIndirection(int32_t* tgt_indir_cache,
+                                                           const int32_t* src_indir_cache,
+                                                           const int32_t* beam_ids,
+                                                           int batch_size,
+                                                           int beam_width,
+                                                           int input_seq_length,
+                                                           int max_seq_length,
+                                                           int current_length,
+                                                           cudaStream_t stream) {
+  const dim3 block(32);
```

Comment:
Is this block size optimized for common scenario?

Reply:
It seems to be. From the profiling logs for some inputs I tried, this kernel is always very cheap - 0.1%.  If we notice perf issues in future because of this, we can tune this further. Also this is the same block size used by FT.